### DISCUSSION TITLE
Create Wifi Hotspot on linux operating system

### DISCUSSION THREAD

**Discussion by opeolluwa** [opeolluwa/filesync#179] [Ideas]
Implement a feature that allows the creation of a Wifi Hotspot on the Linux operating system in the [linux_hotspot.rs](https://github.com/opeolluwa/send-file/blob/master/core/src/utils/linux_hotspot.rs) module. The hotspot must be password protected. 

The following projects might be helpful 
- [https://github.com/lakinduakash/linux-wifi-hotspot](https://github.com/lakinduakash/linux-wifi-hotspot)
- [https://github.com/balena-os/wifi-connect](https://github.com/balena-os/wifi-connect)

**Comment by 0xkelvin** [comment]
hello let assign to me, i would like to work on this issue

**Comment by opeolluwa** [comment]
Hi, @0xkelvin thank you for reaching out to contribute 😊   
I appreciate it ❤️. Please read the [contribution guide](https://github.com/opeolluwa/send-file/blob/master/CONTRIBUTING.md) and [code of conduct](https://github.com/opeolluwa/send-file/blob/master/CODE_OF_CONDUCT.md) to get started

**Comment by opeolluwa** [comment]
> Hello let assign to me, i would like to work on this issue

done! 😊 
Please note that the feature must be callable as a function, it will also be great if the code is well commented and unit tests, are written for ease of maintenance in the future.

Again I appreciate you reaching out to contribute 💯 
Let me know if you have any questions

**Comment by opeolluwa** [comment]
Hi, @0xkelvin how is it going?

**Comment by 0xkelvin** [comment]
hello, learning a bit your source coding, i am a bit slow

**Comment by opeolluwa** [comment]
It's fine. 
Please take your time. I was only hoping to check the state of things with you

**Comment by opeolluwa** [comment]
Hi @0xkelvin, hope you've figured out the project structure? Please, Let me know if you need anything

**Comment by opeolluwa** [comment]
I found this [medium article ](https://anooppoommen.medium.com/create-a-wifi-hotspot-on-linux-29349b9c582d) and a corresponding implementation in https://github.com/lakinduakash/linux-wifi-hotspot/tree/master/src/scripts. The script is currently extracted to the [core/src/scripts/Linux-hotspot path
](https://github.com/opeolluwa/send-file/tree/master/core/src/scripts/linux-hotspot)

The challenge here is that the script requires superuser privilege to work (sudo), which might be bad for the user experience, it would be great if there is a way to bypass this and then call the script from the existing Rust code using Rust [std::process::Command](https://doc.rust-lang.org/std/process/struct.Command.html)

**Comment by opeolluwa** [comment]
create hotspot using nmcli https://gist.github.com/narate/d3f001c97e1c981a59f94cd76f041140

**Comment by opeolluwa** [comment]
create hotspot using nmcli https://gist.github.com/narate/d3f001c97e1c981a59f94cd76f041140

**Reply by opeolluwa** [reply]
a post on stack exchange which address this https://unix.stackexchange.com/questions/234552/create-wireless-access-point-and-share-internet-connection-with-nmcli/310699#310699

**Reply by opeolluwa** [reply]
For this to be effective, the Wifi interface is required, [this post](https://unix.stackexchange.com/questions/286721/get-wi-fi-interface-device-names) addresses, getting WiFi interface

**Comment by opeolluwa** [comment]
A shell script to create hotspot for Linux, extracts [from](https://gitlab.com/dakataca/dotfiles/-/blob/dev/files/networkmanager/hostpot.sh)

```bash
#!/bin/bash

# Script para crear wifi hostpot a partir de la red cableada.

output="$(nmcli d s)"

IFNAME=$(ls /sys/class/net/ | grep "^wl.\+")
SSID="MATARIFE"
PASSWD="p4r4qu1+0m4+4r1f3"


# Crear perfil de red.
create_wifi_profile() {

	nmcli c add type wifi ifname $IFNAME con-name $SSID autoconnect no ssid $SSID 802-11-wireless.mode ap 802-11-wireless.band bg ipv4.method shared wifi-sec.key-mgmt wpa-psk wifi-sec.psk $PASSWD 

}


hostpot(){

	if [[ `nmcli con show | grep -ow "$SSID"` ]]; then
		
		echo "Perfil de red \"$SSID\" ya existe."
		nmcli con up MATARIFE 
	else
		echo "Perfil de red \"$SSID\" aun no existe."
		create_wifi_profile
		sleep 1
		nmcli con up $SSID 
	fi

}


switch_hostpot(){

	# Si el wifi está conectado.
	if [[ `echo "$output" | awk '/^wl/{print $3}'` = "conectado" ]]; then 
 
		# Deshabilitar conexión activa con nmcli.
		#nmcli con down "$(nmcli d s | awk '/^wl/{for (i=1; i<=NF-3; i++) $i = $(i+3); NF-=3; print}')"
		hostpot
	else
		echo "No hacer nada."
		nmcli d s
		hostpot
	fi

}


# Si el cable de red está conectado.
if [[ `echo "$output" | awk '/^en/{print $3}'` = "conectado" ]]; then 

	echo "Cable de red conectado."

	if [[ `echo "$output" | awk '/^wl/{for (i=1; i<=NF-2; i++) $i = $(i+2); NF-=3; print}'` = "no disponible" ]]; then 

		#Enciendo el dispositivo inalámbrico con la utilidad nmcli.
		nmcli radio wifi on && sleep 4
		switch_hostpot
	else
		switch_hostpot
	fi
else
	echo "Cable de red desconectado."
fi
```

**Comment by opeolluwa** [comment]
see these
https://xy2401.com/local-docs/redhat/8/configuring-and-managing-networking/OEBPS/Configuring-Networking-with-nmcli_configuring-and-managing-networking.html#tabl-Aliases-and-property-names
https://developer-old.gnome.org/NetworkManager/stable/nmcli.html

**Comment by opeolluwa** [comment]
See this on Windows hotspot
https://github.com/JamesCullum/Windows-Hotspot
